commons/src/main/java/org/infinispan/multimap/api/BasicMultimapCache.java

```diff
@@ -0,0 +1,184 @@
+package org.infinispan.multimap.api;
+
+import java.util.Collection;
+import java.util.concurrent.CompletableFuture;
+
+import org.infinispan.commons.util.Experimental;
+
+/**
+ * {@link BasicMultimapCache} provides the common building block for the two different types of multimap caches that Infinispan
+ * provides: embedded and remote. <p> Please see the <a href="http://infinispan.org/documentation/">Infinispan
+ * documentation</a> and/or the <a href="http://infinispan.org/docs/dev/getting_started/getting_started.html">5 Minute
+ * Usage Tutorial</a> for more details on Infinispan.
+ * <p/>
+ * <p> MutimapCache is a type of Infinispan Cache that maps keys to values, similar to {@link
+ * org.infinispan.commons.api.AsyncCache} in which each key can contain multiple values.
+ * <pre>
+ *    foo -> 1
+ *    bar -> 3, 4, 5
+ * </pre>
+ * <p> <h2>Example</h2>
+ * <pre>
+ *
+ *    multimapCache.put("k", "v1").join();
+ *    multimapCache.put("k", "v2").join();
+ *    multimapCache.put("k", "v3").join();
+ *
+ *    Collection<String> results = multimapCache.get("k").join();
+ *
+ * </pre>
+ * <p> <h2>Eviction</h2> <p> Eviction works per key. This means all the values associated on a key will be evicted.
+ * </p>
+ * <p>
+ * <h2>Views</h2>
+ * <p>
+ * The returned collections when calling "get" are views of the values on the key. Any change on these collections won't
+ * affect the cache values on the key.
+ * <p>
+ * <h2>Null values</h2> Null values are not supported. The multimap cache won't have a null key or any null value.
+ * <p>
+ * Example
+ * <pre>
+ *     multimapCache.put(null, "v1").join() -> fails
+ *     multimapCache.put("k", null).join() -> fails
+ *     multimapCache.put("k", "v1").join() -> works and add's v1
+ *     multimapCache.containsKey("k").join() -> true
+ *     multimapCache.remove("k", "v1").join() -> works, removes v1 and as the remaining collection is empty, the key is
+ * removed
+ *     multimapCache.containsKey("k").join() -> false
+ *  </pre>
+ * <p>
+ * <p>
+ * <h2>Duplicates</h2> The current implementation does not support duplicate values on keys. {@link
+ * Object#equals(Object)} method is used to check if a value is already present in the key. This means that the
+ * following code
+ * <pre>
+ *    multimapCache.put("k", "v1").join();
+ *    multimapCache.put("k", "v2").join();
+ *    multimapCache.put("k", "v2").join();
+ *    multimapCache.put("k", "v2").join();
+ *
+ *    multimapCache.get("k").thenAccept(values -> System.out.println(values.size()));
+ * </pre>
+ * prints the value 2. "k" -> ["v1", "v"]
+ * <p>
+ * Future implementations might evolve to support duplicated values.
+ * <p>
+ * <h2>Transactions</h2> MultimapCache supports implicit transactions without blocking. The following methods block when
```

Comment:
ATM transactions are specific to embedded mode. Also, all Hot Rod operations are blocking.

Reply:
I think I have to review the doc of this super interface moved to commons